doc(ts): Add typescript documentation (#2463)
codeceptjs/CodeceptJS: docs/typescript.md

---
permalink: /typescript
title: TypeScript
---

# Typescript

> CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS

Reviewer: ```suggestion
CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS.
```

> CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means now you can write your tests in TS. Also, all of your custom steps could be written in TS

Reviewer 2: ```suggestion
CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that you can write your tests in TS. Also, all of your custom steps can be written in TS
```

> CodeceptJS supports [type declaration](https://github.com/codecept-js/CodeceptJS/tree/master/typings) for [TypeScript](https://www.typescriptlang.org/). It means that allow you to write your tests in TypeScript. Also, all of your custom steps could be write in TypeScript.

Reviewer 3: It means now you can write your tests in TS. Also, all of your custom steps could be written in TS

# Why TypeScript?

> With the TypeScript writing CodeceptJS tests becomes much easier. If you configure TS properly in your project as well as your IDE, you will get the following features:

Reviewer 2: don't use word "easy" in documentation

> With the TypeScript writing CodeceptJS tests becomes much easier. If you configure TS properly in your project as well as your IDE, you will get the following features:

Reviewer: ```suggestion
If you configure TS properly in your project as well as your IDE, you will get the following features:
```

> With the TypeScript writing CodeceptJS tests becomes much easier. If correctly configured TypeScript in your project and correctly configured your IDE, then you can get the following features as result:

Reviewer 3: If you configure TS properly in your project as well as your IDE, you will get the following features:

> - [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.

Reviewer: @Vorobeyko "IntelliSense" instead of "InteleSence"

> - [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.

Reviewer: ```suggestion
- [Autocomplete (with IntelliSense)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.
```

> - [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that tries complete your thoughts for you and suggest you which function or property exist in class, what arguments can be passed to the method, what method return etc. 

Reviewer 3: I’m not sure InteleSence here is a tool or something? 😅

> - [Autocomplete (with InteleSence)](https://code.visualstudio.com/docs/editor/intellisense) - a tool that tries complete your thoughts for you and suggest you which function or property exist in class, what arguments can be passed to the method, what method return etc. 

Reviewer 3: A tool that streamlines your work by suggesting when you typing what function or property which exists in a class, what arguments can be passed to that method, what it returns, etc.

Example:

![Auto Complete](/img/Auto_comlete.gif)

> - Quick info about members - show info about each of your steps right in your tests. Example:

Reviewer 2: ```suggestion
- To show additional information for a step in a test. Example:
```

> - Quick info about members - show info about each of your steps right in your tests. Example:

Reviewer 2: probably it was items of a list

![Quick Info](/img/Quick_info.gif)

- Checks types - thanks to TypeScript support in CodeceptJS now allow to tests your tests. TypeScript can prevent some errors: 
  - invalid type of variables passed to function;
  - calls no-exist method from PageObject or `I` object;
  - incorrect used CodeceptJS features; 

> 

Reviewer: ```suggestion
```

## Getting Started

> ### Instal TypeScipt

Reviewer 2: ```suggestion
### Install TypeScipt
```

> For writing tests in TypeScript you`ll need to install TypeScript and ts-node in your project.

Reviewer 2: ```suggestion
For writing tests in TypeScript you`ll need to install `typescript` and `ts-node` into your project.
```

> `npm install typescript ts-node`

Reviewer: ```suggestion
`npm install typescript ts-node -S`
```

### Configure codecept.conf.js

> To configure TypeScript in your project, you need to add [`ts-node/register`](https://github.com/TypeStrong/ts-node) on first line in your config. Like in the following config file:

Reviewer: ```suggestion
To configure TypeScript in your project, you need to require [`ts-node/register`](https://github.com/TypeStrong/ts-node) on first line in your config. Like in the following config file:
```

```js
require('ts-node/register')

exports.config = {
  tests: './*_test.ts',
  output: './output',
  helpers: {
    Puppeteer: {
      url: 'http://example.com',
    },
  },
  name: 'project name',
}
```

### Configure tsconfig.json

We recommended the following configuration in a [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html):

```json
{
  "ts-node": {
    "files": true
  },
  "compilerOptions": {
    "target": "es2018",
    "lib": ["es2018", "DOM"],
    "esModuleInterop": true,
    "module": "commonjs",
    "strictNullChecks": true,
    "types": ["codeceptjs"],
  },
}
```

> You can find an example project with TypeScript and CodeceptJS on our project [typescript-boilerplate](https://github.com/codecept-js/typescript-boilerplate).

> ### Set up steps.d.ts

Reviewer 2: ```suggestion
### Set Up steps.d.ts
```

> Configuring the `tsconfig.json` and `codecept.conf.js` is not enough, you will need to configure the `steps.d.ts` file for custom steps. Just simply do this by running this command::

Reviewer: ```suggestion
Configuring the `tsconfig.json` and `codecept.conf.js` is not enough, you will need to configure the `steps.d.ts` file for custom steps. Just simply do this by running this command:
```

> Not enough to configure `tsconfig.json` and `codecept.cong.js`. Also need to configure `steps.d.ts` file with custom steps. For it need to run def command:

Reviewer 3: Configuring the tsconfig.json and codecept.conf.js is not enough, you will need to configure the step.d.ts file for custom steps. Just simply do this by running this command:

`npx codeceptjs def`

As a result, a file will be created on your root folder with following content:

```ts
/// <reference types='codeceptjs' />

declare namespace CodeceptJS {
  interface SupportObject { I: I }
  interface Methods extends Puppeteer {}
  interface I extends WithTranslation<Methods> {}
  namespace Translation {
    interface Actions {}
  }
}

```

> ## Types for custom helper or page object

Reviewer 2: please have separate chapters for helper & for page object

> If you want to gets types for your [custom helper](https://codecept.io/helpers/#configuration), you can add their automatically with CodeceptJS command `npx codeceptjs def`. 

Reviewer 2: ```suggestion
If you want to get types for your [custom helper](https://codecept.io/helpers/#configuration), you can add their automatically with CodeceptJS command `npx codeceptjs def`. 
```

For example, if you add the new step `printMessage` for your custom helper like this:
```js
// customHelper.ts
class CustomHelper extends Helper {
  printMessage(msg: string) {
    console.log(msg)
  }
}

export = CustomHelper
```

Then you need to add this helper to your `codecept.conf.js` like in this [docs](https://codecept.io/helpers/#configuration).

> And then run to command `npx codeceptjs def`. 

Reviewer 2: ```suggestion
And then run the command `npx codeceptjs def`. 
```

As result our `steps.d.ts` file will be updated like this: 
```ts
/// <reference types='codeceptjs' />
type CustomHelper = import('./CustomHelper');

declare namespace CodeceptJS {
  interface SupportObject { I: I }
  interface Methods extends Puppeteer, CustomHelper {}
  interface I extends WithTranslation<Methods> {}
  namespace Translation {
    interface Actions {}
  }
}
```

> And now you can use autocomplete on your test.

Reviewer: ```suggestion
```

> Generation types for PageObject looks like for a custom helper, but `steps.d.ts` will look like:

Reviewer: @Vorobeyko what do you mean by "Generation types"?

Author: Maybe - "To generate types", sorry

> Generation types for PageObject looks like for a custom helper, but `steps.d.ts` will look like:

Reviewer: ```suggestion
Using `npx codeceptjs def` will also generate types for PageObjects. `steps.d.ts` will look like:
```

> Generation types for PageObject looks like for custom helper, but `steps.d.ts` will be looks like:

Reviewer 2: ```suggestion
Generation types for PageObject looks like for a custom helper, but `steps.d.ts` will look like:
```
